File: generation/sam_snap.py
#!/usr/bin/env python3
"""SUPERIOR detection pipeline: prior-guided SAM-3.1 snap.

For each skin, every control's template rect is fed to SAM 3.1 as a BOX PROMPT
(slightly padded to tolerate drift). SAM segments the actual painted control at
that spot and returns a tight, normalized box + a confidence score — robust to
material (works on a dark socket OR a bright painted knob), unlike darkness
thresholding. We then:
  - snap the rect to SAM's box when confidence is high AND the move is plausible
  - keep the template prior when SAM is unsure (never snap to garbage)
  - apply STRUCTURAL constraints to grouped controls (the 5-button row and the
    6-band EQ are evenly spaced; refit the group so a missed/merged member is
    interpolated from its neighbours rather than left wrong)

Outputs a `<id>.snap.json` sidecar (NOT overwriting template.json) + an overlay
PNG comparing prior (red) vs snapped (green) so the result can be reviewed before
it's applied.

Usage: python3 sam_snap.py <id> [<id> ...]
"""
import json, os, sys, time, urllib.request
import numpy as np
from PIL import Image, ImageDraw, ImageFont

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SK = os.path.join(ROOT, "public", "skins")
OUT = "/Users/conner/Desktop/cc-skeuo"
W, H = 1024, 1536
FAL = os.environ.get("FAL_KEY")
SAM = "https://queue.fal.run/fal-ai/sam-3-1/image"
PAD = 0.18          # enlarge the box prompt by this fraction to catch drift
MIN_SCORE = 0.55    # below this, keep the prior
CTRL = {"button", "toggle", "knob", "slider-h", "slider-v", "slider-arc", "slider-path"}


def _req(url, data=None, headers=None, method=None):
    r = urllib.request.Request(url, data=data, headers=headers or {}, method=method)
    try:
        with urllib.request.urlopen(r, timeout=120) as resp:
            raw = resp.read().decode().strip()
    except urllib.error.HTTPError as e:
        body = e.read().decode()[:500]
        raise SystemExit(f"HTTP {e.code} on {url}\n{body}")
    return json.loads(raw) if raw else {}


def upload(path):
    name = os.path.basename(path)
    init = _req("https://rest.alpha.fal.ai/storage/upload/initiate",
                data=json.dumps({"file_name": name, "content_type": "image/png"}).encode(),
                headers={"Authorization": f"Key {FAL}", "Content-Type": "application/json"})
    with open(path, "rb") as f:
        body = f.read()
    _req(init["upload_url"], data=body, headers={"Content-Type": "image/png"}, method="PUT")
    return init["file_url"]


def composite(skin):
    im = Image.open(os.path.join(SK, skin, "frame.png")).convert("RGBA").resize((W, H))
    bg = Image.new("RGBA", (W, H), (26, 26, 28, 255)); bg.alpha_composite(im)
    p = f"/tmp/sam-{skin}.png"; bg.convert("RGB").save(p); return p


def pad_box(rc):
    px, py = rc["w"]*PAD, rc["h"]*PAD
    x0 = max(0.0, rc["x"]-px); y0 = max(0.0, rc["y"]-py)
    x1 = min(1.0, rc["x"]+rc["w"]+px); y1 = min(1.0, rc["y"]+rc["h"]+py)
    return [round(x0*W), round(y0*H), round(x1*W), round(y1*H)]


def detect(url, regs):
    """One SAM call with a box prompt per control. Returns list aligned to ctrl regs:
    {cx,cy,w,h,score} (normalized) or None."""
    ctrl = [r for r in regs if r["kind"] in CTRL]
    boxes = [pad_box(r["rect"]) for r in ctrl]
    payload = {"image_url": url,
               "box_prompts": [{"x_min": b[0], "y_min": b[1], "x_max": b[2], "y_max": b[3]} for b in boxes],
               "include_boxes": True, "include_scores": True,
               "return_multiple_masks": True, "max_masks": len(boxes), "apply_mask": False}
    sub = _req(SAM, data=json.dumps(payload).encode(),
               headers={"Authorization": f"Key {FAL}", "Content-Type": "application/json"})
    t0 = time.time()
    while True:
        st = _req(sub["status_url"], headers={"Authorization": f"Key {FAL}"})
        if st.get("status") == "COMPLETED": break
        if st.get("status") in ("FAILED", "ERROR") or time.time()-t0 > 120:
            raise SystemExit(f"SAM failed: {st}")
        time.sleep(1.5)
    res = _req(sub["response_url"], headers={"Authorization": f"Key {FAL}"})
    # metadata carries index→box mapping; align back to ctrl order
    out = [None]*len(ctrl)
    for m in res.get("metadata") or []:
        i = m.get("index")
        if i is None or not (0 <= i < len(ctrl)):
            continue
        b = m["box"]            # [cx,cy,w,h] normalized
        out[i] = {"cx": b[0], "cy": b[1], "w": b[2], "h": b[3], "score": m.get("score", 0)}
    return ctrl, out


def plausible(rc, d):
    """Reject SAM boxes that wandered or ballooned — keep the prior instead."""
    if d is None or d["score"] < MIN_SCORE:
        return False
    pcx, pcy = rc["x"]+rc["w"]/2, rc["y"]+rc["h"]/2
    move = ((d["cx"]-pcx)**2 + (d["cy"]-pcy)**2) ** 0.5
    if move > max(rc["w"], rc["h"]) * 1.2:        # moved more than ~one control away
        return False
    aw, ah = d["w"]/rc["w"], d["h"]/rc["h"]
    if not (0.4 < aw < 2.6 and 0.4 < ah < 2.6):   # size sanity
        return False
    return True


def snap(ctrl, det):
    """Per-control snap with fallback, then structural refit of grouped rows."""
    new = []
    flags = []
    for r, d in zip(ctrl, det):
        rc = r["rect"]
        if plausible(rc, d):
            nx = {"x": d["cx"]-d["w"]/2, "y": d["cy"]-d["h"]/2, "w": d["w"], "h": d["h"]}
            new.append(nx); flags.append(("snap", d["score"]))
        else:
            new.append(dict(rc)); flags.append(("keep", d["score"] if d else 0))

    # structural refit: evenly-spaced rows (eqBand×N, the 5 transport buttons)
    def refit_row(pred):
        idx = [i for i, r in enumerate(ctrl) if pred(r)]
        if len(idx) < 3: return
        snapped = [i for i in idx if flags[i][0] == "snap"]
        if len(snapped) < 2: return
        # robust line through snapped centers vs their order; reapply to all in row
        order = sorted(idx, key=lambda i: ctrl[i]["rect"]["x"])
        xs = np.array([new[i]["x"]+new[i]["w"]/2 for i in order])
        pos = np.arange(len(order))
        good = np.array([flags[i][0] == "snap" for i in order])
        if good.sum() < 2: return
        a, b = np.polyfit(pos[good], xs[good], 1)         # cx ≈ a*pos + b
        good_idx = [order[k] for k in range(len(order)) if good[k]]
        cy = float(np.median([new[i]["y"]+new[i]["h"]/2 for i in good_idx]))
        wmd = float(np.median([new[i]["w"] for i in good_idx]))
        hmd = float(np.median([new[i]["h"] for i in good_idx]))
        for k, i in enumerate(order):
            if flags[i][0] == "snap":
                continue
            cx = a*k + b
            new[i] = {"x": cx-wmd/2, "y": cy-hmd/2, "w": wmd, "h": hmd}
            flags[i] = ("refit", flags[i][1])

    refit_row(lambda r: r.get("bind") == "eqBand")
    refit_row(lambda r: r["kind"] == "button" and r.get("bind") in ("prev","play","pause","stop","next"))
    return new, flags


def font(s):
    try: return ImageFont.truetype("/System/Library/Fonts/Supplemental/Arial Bold.ttf", s)
    except: return ImageFont.load_default()


def process(skin, apply=False):
    tpl = json.load(open(os.path.join(SK, skin, "template.json")))
    regs = tpl["regions"]
    url = upload(composite(skin))
    ctrl, det = detect(url, regs)
    new, flags = snap(ctrl, det)

    # overlay: prior (red) vs snapped (green/amber)
    fr = Image.open(os.path.join(SK, skin, "frame.png")).convert("RGBA").resize((W, H))
    bg = Image.new("RGBA", (W, H), (22, 22, 24, 255)); bg.alpha_composite(fr)
    im = bg.convert("RGB"); g = ImageDraw.Draw(im); f = font(20); lw = 4
    COL = {"snap": (90, 255, 120), "refit": (255, 190, 60), "keep": (120, 120, 130)}
    moves = []
    for r, nx, fl in zip(ctrl, new, flags):
        rc = r["rect"]
        g.rectangle([rc["x"]*W, rc["y"]*H, (rc["x"]+rc["w"])*W, (rc["y"]+rc["h"])*H],
                    outline=(255, 70, 70), width=2)                       # prior = red
        c = COL[fl[0]]
        g.rectangle([nx["x"]*W, nx["y"]*H, (nx["x"]+nx["w"])*W, (nx["y"]+nx["h"])*H],
                    outline=c, width=lw)                                  # snapped = green
        mv = (((rc["x"]+rc["w"]/2)-(nx["x"]+nx["w"]/2))**2 + ((rc["y"]+rc["h"]/2)-(nx["y"]+nx["h"]/2))**2)**0.5
        moves.append((r.get("bind") or r["id"], fl[0], round(fl[1], 2), round(mv*1000)))
    op = os.path.join(OUT, f"2026-06-16-snap-{skin}.png"); im.save(op)

    # write sidecar (or apply)
    snapped_regs = []
    ci = 0
    for r in regs:
        r2 = dict(r)
        if r["kind"] in CTRL:
            r2["rect"] = {k: round(v, 4) for k, v in new[ci].items()}; ci += 1
        snapped_regs.append(r2)
    sidecar = dict(tpl); sidecar["regions"] = snapped_regs
    sp = os.path.join(SK, skin, "template.snap.json")
    json.dump(sidecar, open(sp, "w"), indent=2)
    if apply:
        json.dump(sidecar, open(os.path.join(SK, skin, "template.json"), "w"), indent=2)
    n_snap = sum(1 for _, s, _, _ in moves if s == "snap")
    print(f"[{skin}] {n_snap}/{len(moves)} snapped  overlay={op}")
    for name, st, sc, mv in moves:
        print(f"    {name:9} {st:6} conf {sc:<5} moved {mv}‰")
    return op


if __name__ == "__main__":
    if not FAL: raise SystemExit("FAL_KEY not in env")
    apply = "--apply" in sys.argv
    ids = [a for a in sys.argv[1:] if not a.startswith("--")]
    for sid in ids:
        process(sid, apply=apply)

File: src/generate/samSnap.ts
// Prior-guided SAM-3.1 snap — the ALIGN pass, ported from generation/sam_snap.py.
//
// WHY: the painter places controls/screens NEAR the blueprint sockets but not pixel-on,
// so the template rects (used to overlay live controls + render the screen/visualizer)
// land off the painted features → "everything outside bounds". Eyeballing or darkness
// heuristics are unreliable; SAM-3.1 is a real segmentation model. Each control's rect is
// fed to SAM as a (gap-clamped) BOX PROMPT; SAM segments the actual painted control there
// and returns a tight normalized box + score. We snap when confident + the move is
// plausible, keep the prior when unsure, and structurally refit evenly-spaced rows
// (transport buttons, EQ bands). Robust on a dark socket OR a bright painted knob.
//
// One SAM call per skin (all box prompts batched). Server-side only (FAL_KEY).

const SAM_URL = "https://queue.fal.run/fal-ai/sam-3-1/image";
const PAD = 0.6;        // expand the box prompt to tolerate drift…
const PAD_GAP = 0.45;   // …but never past this fraction of the gap to a neighbour (dense rows)
const MIN_SCORE = 0.55; // below this, keep the prior
const SAM_W = 1024, SAM_H = 1536; // device aspect (2:3); box prompts are in these pixels

// Snappable kinds: controls AND displays. Including "display" is the key for the
// visualizer/marquee alignment — SAM box-prompts the painted screen glass and snaps the
// region onto it, so live content stops landing off the actual screen.
const CTRL = new Set(["button", "toggle", "knob", "slider-h", "slider-v", "slider-arc", "slider-path", "display"]);

interface Rect { x: number; y: number; w: number; h: number }
interface Region { kind: string; bind?: string; id?: string; rect: Rect; [k: string]: unknown }
interface Det { cx: number; cy: number; w: number; h: number; score: number }

// per-control pad: PAD×size, clamped to PAD_GAP×(gap to nearest neighbour) so dense rows
// (EQ) don't grow into each other and get merged by SAM.
function adaptivePads(rects: Rect[]): [number, number][] {
  return rects.map((a, i) => {
    let gx = 1.0, gy = 1.0;
    rects.forEach((b, j) => {
      if (i === j) return;
      if (!(a.y + a.h < b.y || b.y + b.h < a.y)) { // shares a row
        const gap = Math.abs((b.x + b.w / 2) - (a.x + a.w / 2)) - (a.w + b.w) / 2;
        if (gap >= -0.01) gx = Math.min(gx, Math.max(0, gap));
      }
      if (!(a.x + a.w < b.x || b.x + b.w < a.x)) { // shares a column
        const gap = Math.abs((b.y + b.h / 2) - (a.y + a.h / 2)) - (a.h + b.h) / 2;
        if (gap >= -0.01) gy = Math.min(gy, Math.max(0, gap));
      }
    });
    return [Math.min(a.w * PAD, PAD_GAP * gx), Math.min(a.h * PAD, PAD_GAP * gy)];
  });
}

// box prompt in the IMAGE's ACTUAL pixel space (SAM interprets box_prompts as image
// pixels — using a fixed size when the upload is a different size makes every prompt land
// in the wrong place → nothing detected).
function padBox(rc: Rect, [px, py]: [number, number], W: number, H: number): [number, number, number, number] {
  const x0 = Math.max(0, rc.x - px), y0 = Math.max(0, rc.y - py);
  const x1 = Math.min(1, rc.x + rc.w + px), y1 = Math.min(1, rc.y + rc.h + py);
  return [Math.round(x0 * W), Math.round(y0 * H), Math.round(x1 * W), Math.round(y1 * H)];
}

// PNG dimensions from the IHDR header (bytes 16–24), no decode.
function pngDims(b: Uint8Array): { w: number; h: number } | null {
  if (b.length < 24 || b[0] !== 0x89 || b[1] !== 0x50) return null;
  const rd = (o: number) => ((b[o] << 24) | (b[o + 1] << 16) | (b[o + 2] << 8) | b[o + 3]) >>> 0;
  return { w: rd(16), h: rd(20) };
}

function plausible(rc: Rect, d: Det | null): boolean {
  if (!d || d.score < MIN_SCORE) return false;
  const pcx = rc.x + rc.w / 2, pcy = rc.y + rc.h / 2;
  const move = Math.hypot(d.cx - pcx, d.cy - pcy);
  if (move > Math.max(rc.w, rc.h) * 1.2) return false;     // wandered more than ~one control away
  const aw = d.w / rc.w, ah = d.h / rc.h;
  return aw > 0.4 && aw < 2.6 && ah > 0.4 && ah < 2.6;     // size sanity
}

// Upload image bytes to fal storage → a fal-reachable URL (needed when the source isn't
// publicly fetchable by fal, e.g. a localhost /api/asset in dev). Mirrors sam_snap.py upload().
export async function uploadToFal(falKey: string, bytes: Uint8Array): Promise<string> {
  const init = await fetchJson("https://rest.alpha.fal.ai/storage/upload/initiate", {
    method: "POST",
    headers: { Authorization: `Key ${falKey}`, "Content-Type": "application/json" },
    body: JSON.stringify({ file_name: "snap.png", content_type: "image/png" }),
  });
  await fetch(init.upload_url as string, { method: "PUT", headers: { "Content-Type": "image/png" }, body: bytes as unknown as ArrayBuffer });
  return init.file_url as string;
}

async function fetchJson(url: string, init?: RequestInit): Promise<Record<string, unknown>> {
  const r = await fetch(url, init);
  if (!r.ok) throw new Error(`SAM ${r.status}: ${(await r.text()).slice(0, 300)}`);
  const t = (await r.text()).trim();
  return t ? JSON.parse(t) : {};
}

// one SAM call: a box prompt per control. Returns dets aligned to the ctrl order.
async function detect(falKey: string, imageUrl: string, ctrl: Region[], W: number, H: number): Promise<(Det | null)[]> {
  const pads = adaptivePads(ctrl.map((r) => r.rect));
  const boxes = ctrl.map((r, i) => padBox(r.rect, pads[i], W, H));
  const payload = {
    image_url: imageUrl,
    box_prompts: boxes.map((b) => ({ x_min: b[0], y_min: b[1], x_max: b[2], y_max: b[3] })),
    include_boxes: true, include_scores: true,
    return_multiple_masks: true, max_masks: boxes.length, apply_mask: false,
  };
  const auth = { Authorization: `Key ${falKey}`, "Content-Type": "application/json" };
  const sub = await fetchJson(SAM_URL, { method: "POST", headers: auth, body: JSON.stringify(payload) });
  const statusUrl = sub.status_url as string, responseUrl = sub.response_url as string;
  const t0 = Date.now();
  for (;;) {
    const st = await fetchJson(statusUrl, { headers: { Authorization: `Key ${falKey}` } });
    if (st.status === "COMPLETED") break;
    if (st.status === "FAILED" || st.status === "ERROR" || Date.now() - t0 > 120_000) throw new Error(`SAM status ${st.status}`);
    await new Promise((r) => setTimeout(r, 1500));
  }
  const res = await fetchJson(responseUrl, { headers: { Authorization: `Key ${falKey}` } });
  const out: (Det | null)[] = new Array(ctrl.length).fill(null);
  for (const m of (res.metadata as Array<Record<string, unknown>>) ?? []) {
    const i = m.index as number;
    if (i == null || i < 0 || i >= ctrl.length) continue;
    const b = m.box as number[]; // [cx,cy,w,h] normalized
    out[i] = { cx: b[0], cy: b[1], w: b[2], h: b[3], score: (m.score as number) ?? 0 };
  }
  return out;
}

// least-squares slope/intercept for cx ≈ a*pos + b over the "good" (snapped) members
function lineFit(pos: number[], xs: number[]): [number, number] {
  const n = pos.length, sx = pos.reduce((s, v) => s + v, 0), sy = xs.reduce((s, v) => s + v, 0);
  const sxx = pos.reduce((s, v) => s + v * v, 0), sxy = pos.reduce((s, v, i) => s + v * xs[i], 0);
  const d = n * sxx - sx * sx || 1;
  const a = (n * sxy - sx * sy) / d;
  return [a, (sy - a * sx) / n];
}
const median = (a: number[]) => { const s = [...a].sort((x, y) => x - y); const m = s.length >> 1; return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2; };

// Snap each control's rect to SAM (when confident), then structurally refit
// evenly-spaced rows (transport buttons, EQ bands) so a missed member is interpolated.
// Returns a NEW regions array with corrected rects (only CTRL kinds change).
// snap controls, and ONLY the live screen among displays (visualizer/cd/albumart) — NOT
// small text displays (marquee/time/playlist), whose box prompts overlap neighbours and
// mis-snap onto a button; their layout position is reliable, so keep it.
const BIG_SCREEN = new Set(["visualizer", "cd", "albumart"]);
function snappable(r: Region): boolean {
  if (!CTRL.has(r.kind)) return false;
  if (r.kind === "display") return BIG_SCREEN.has(String((r as { dynamicType?: string }).dynamicType ?? ""));
  return true;
}

export async function snapRegions(falKey: string, imageUrl: string, regions: Region[], imgW = SAM_W, imgH = SAM_H): Promise<{ regions: Region[]; snapped: number; total: number }> {
  const ctrl = regions.filter(snappable);
  if (!ctrl.length) return { regions, snapped: 0, total: 0 };
  const det = await detect(falKey, imageUrl, ctrl, imgW, imgH);
  const newRects: Rect[] = [];
  const flags: ("snap" | "keep" | "refit")[] = [];
  ctrl.forEach((r, i) => {
    const d = det[i];
    if (plausible(r.rect, d) && d) {
      // POSITION-ONLY: move the rect's CENTER onto SAM's detection, but KEEP the prior
      // size. Never let detection resize a control — that made knobs/buttons different
      // sizes (ai-image-coords-rule). The blueprint sizes are the load-bearing truth.
      newRects.push({ x: d.cx - r.rect.w / 2, y: d.cy - r.rect.h / 2, w: r.rect.w, h: r.rect.h }); flags.push("snap");
    } else { newRects.push({ ...r.rect }); flags.push("keep"); }
  });

  const refitRow = (pred: (r: Region) => boolean) => {
    const idx = ctrl.map((r, i) => [r, i] as const).filter(([r]) => pred(r)).map(([, i]) => i);
    if (idx.length < 3) return;
    const order = [...idx].sort((a, b) => ctrl[a].rect.x - ctrl[b].rect.x);
    const good = order.map((i) => flags[i] === "snap");
    if (good.filter(Boolean).length < 2) return;
    const gPos: number[] = [], gXs: number[] = [];
    order.forEach((i, k) => { if (good[k]) { gPos.push(k); gXs.push(newRects[i].x + newRects[i].w / 2); } });
    const [a, b] = lineFit(gPos, gXs);
    const gi = order.filter((_, k) => good[k]);
    const cyMed = median(gi.map((i) => newRects[i].y + newRects[i].h / 2));
    order.forEach((i, k) => {
      if (flags[i] === "snap") return;
      // POSITION-ONLY refit: interpolate the row position; KEEP this control's prior size.
      const cx = a * k + b, pw = ctrl[i].rect.w, ph = ctrl[i].rect.h;
      newRects[i] = { x: cx - pw / 2, y: cyMed - ph / 2, w: pw, h: ph };
      flags[i] = "refit";
    });
  };
  refitRow((r) => r.bind === "eqBand");
  refitRow((r) => r.kind === "button" && ["prev", "play", "pause", "stop", "next"].includes(r.bind ?? ""));

  let ci = 0;
  const out = regions.map((r) => {
    if (!snappable(r)) return r;
    const nr = newRects[ci++];
    return { ...r, rect: { x: +nr.x.toFixed(4), y: +nr.y.toFixed(4), w: +nr.w.toFixed(4), h: +nr.h.toFixed(4) } };
  });
  return { regions: out, snapped: flags.filter((f) => f === "snap").length, total: ctrl.length };
}

// Full align from a SERVER-FETCHABLE image URL: fetch the bytes, read true dims, upload to
// fal storage (fal can't fetch localhost; even in prod this is robust), then snap. Used by
// both /api/snap (CF) and the dev plugin so the logic lives in one place.
export async function snapFromUrl(falKey: string, fetchUrl: string, regions: Region[]): Promise<{ regions: Region[]; snapped: number; total: number }> {
  const r = await fetch(fetchUrl);
  if (!r.ok) throw new Error(`fetch image ${r.status}`);
  const bytes = new Uint8Array(await r.arrayBuffer());
  const dims = pngDims(bytes) ?? { w: SAM_W, h: SAM_H };
  const falUrl = await uploadToFal(falKey, bytes);
  return snapRegions(falKey, falUrl, regions, dims.w, dims.h);
}
